src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+                TryMergeLastTwoNodes(contents);
+            }
+
+            if (closeBracketToken.IsMissing)
+            {
+                closeBracketToken = closeBracketToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    WorkspacesResources.Unterminated_character_class_set,
+                    GetTokenStartPositionSpan(_currentToken)));
+            }
+
+            return caretToken.IsMissing
+                ? (RegexBaseCharacterClassNode)new RegexCharacterClassNode(openBracketToken, new RegexSequenceNode(contents.ToImmutableAndFree()), closeBracketToken)
+                : new RegexNegatedCharacterClassNode(openBracketToken, caretToken, new RegexSequenceNode(contents.ToImmutableAndFree()), closeBracketToken);
+        }
+
+        private void ParseCharacterClassComponents(ArrayBuilder<RegexExpressionNode> components)
+        {
+            var left = ParseSingleCharacterClassComponent(isFirst: components.Count == 0, afterRangeMinus: false);
+            if (left.Kind == RegexKind.CharacterClassEscape ||
+                left.Kind == RegexKind.CategoryEscape ||
+                IsEscapedMinus(left))
+            {
+                // \s or \p{Lu} or \- on the left of a minus doesn't start a range. If there is a following
+                // minus, it's just treated textually.
+                components.Add(left);
+                return;
+            }
+
+            if (_currentToken.Kind == RegexKind.MinusToken && !_lexer.IsAt("]"))
+            {
+                // trivia is not allowed anywhere in a character class
+                var minusToken = ConsumeCurrentToken(allowTrivia: false);
+
+                if (_currentToken.Kind == RegexKind.OpenBracketToken)
+                {
+                    components.Add(left);
+                    components.Add(ParseCharacterClassSubtractionNode(minusToken));
+                }
+                else
+                {
+                    var right = ParseRightSideOfCharacterClassRange();
+
+                    if (TryGetRangeComponentValue(left, isRight: false, out var leftCh) &&
+                        TryGetRangeComponentValue(right, isRight: true, out var rightCh) &&
+                        leftCh > rightCh)
+                    {
+                        minusToken = minusToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                            WorkspacesResources.x_y_range_in_reverse_order,
+                            minusToken.GetSpan()));
+                    }
+
+                    components.Add(new RegexCharacterClassRangeNode(left, minusToken, right));
+                }
+            }
+            else
+            {
+                components.Add(left);
+            }
+        }
+
+        private bool IsEscapedMinus(RegexNode node)
+            => node is RegexSimpleEscapeNode simple && IsTextChar(simple.TypeToken, '-');
+
+        private bool TryGetRangeComponentValue(RegexExpressionNode component, bool isRight, out char ch)
```

Comment:
`isRight` is unused. This likely relates to my question on `TryGetRangeComponentValueWorker` logic for sequences.

Reply:
i will investigate what's going on here.

Reply:
See above.  This parameter is unnecessary, and hte logic *is* supposed to only grab the rightmost chunk in a sequence of character-class sub-components.

Reply:
Removed parameter and added large comment explaining why we're doing what we're doing.

Reply:
Consider linking to the corefx issue for this bug.

---
In reply to: [178490903](https://github.com/dotnet/roslyn/pull/23984#discussion_r178490903) [](ancestors = 178490903)